### PR TITLE
Support for Short, Long and Double data types.

### PR DESCRIPTION
**Pull request by PantelisAndrianakis**
Attempt to add support for Short, Long and Double data types.

### CODE REVIEW

**Comment by CodeSpartan** [comment]
Thanks, I'll have a look over the weekend.

File: Source/TcpSocketPlugin/Public/TcpSocketConnection.h [hunk 1]
```diff
@@ -72,34 +72,116 @@ class TCPSOCKETPLUGIN_API ATcpSocketConnection : public AActor
 	UFUNCTION(BlueprintPure, meta = (DisplayName = "Append Bytes", CommutativeAssociativeBinaryOperator = "true"), Category = "Socket")
 	static TArray<uint8> Concat_BytesBytes(TArray<uint8> A, TArray<uint8> B);
 
+	/** Converts a byte to an array of bytes */
+	UFUNCTION(BlueprintPure, meta = (DisplayName = "Byte To Bytes", CompactNodeTitle = "->", Keywords = "cast convert", BlueprintAutocast), Category = "Socket")
+	static TArray<uint8> Conv_ByteToBytes(uint8 InByte);
+
+	/** Converts a short to an array of bytes */
+	UFUNCTION(BlueprintPure, meta = (DisplayName = "Short To Bytes", CompactNodeTitle = "->", Keywords = "cast convert", BlueprintAutocast), Category = "Socket")
+	static TArray<uint8> Conv_ShortToBytes(int16 InShort)
```
Comment (CodeSpartan) [inline Source/TcpSocketPlugin/Public/TcpSocketConnection.h]: Please move the function definition into .cpp
And do in-place initialization, e.g.
```cpp
return TArray<uint8> { 0, 1, 2 };
```

File: Source/TcpSocketPlugin/Public/TcpSocketConnection.h [hunk 2]
```diff
@@ -72,34 +72,116 @@ class TCPSOCKETPLUGIN_API ATcpSocketConnection : public AActor
...
 	/** Converts an integer to an array of bytes */
 	UFUNCTION(BlueprintPure, meta = (DisplayName = "Int To Bytes", CompactNodeTitle = "->", Keywords = "cast convert", BlueprintAutocast), Category = "Socket")
 	static TArray<uint8> Conv_IntToBytes(int32 InInt);
 
-	/** Converts a string to an array of bytes */
-	UFUNCTION(BlueprintPure, meta = (DisplayName = "String To Bytes", CompactNodeTitle = "->", Keywords = "cast convert", BlueprintAutocast), Category = "Socket")
-	static TArray<uint8> Conv_StringToBytes(const FString& InStr);
+	/** Converts a long to an array of bytes */
+	UFUNCTION(BlueprintPure, meta = (DisplayName = "Long To Bytes", CompactNodeTitle = "->", Keywords = "cast convert", BlueprintAutocast), Category = "Socket")
+	static TArray<uint8> Conv_LongToBytes(int64 InLong)
```
Comment (CodeSpartan) [inline Source/TcpSocketPlugin/Public/TcpSocketConnection.h]: Please move the function definition into .cpp
And unroll the loop into an in-place initialization

File: Source/TcpSocketPlugin/Public/TcpSocketConnection.h [hunk 3]
```diff
@@ -72,34 +72,116 @@ class TCPSOCKETPLUGIN_API ATcpSocketConnection : public AActor
...
 	/** Converts a float to an array of bytes */
 	UFUNCTION(BlueprintPure, meta = (DisplayName = "Float To Bytes", CompactNodeTitle = "->", Keywords = "cast convert", BlueprintAutocast), Category = "Socket")
 	static TArray<uint8> Conv_FloatToBytes(float InFloat);
 
-	/** Converts a byte to an array of bytes */
-	UFUNCTION(BlueprintPure, meta = (DisplayName = "Byte To Bytes", CompactNodeTitle = "->", Keywords = "cast convert", BlueprintAutocast), Category = "Socket")
-	static TArray<uint8> Conv_ByteToBytes(uint8 InByte);
+	/** Converts a double to an array of bytes */
+	UFUNCTION(BlueprintPure, meta = (DisplayName = "Double To Bytes", CompactNodeTitle = "->", Keywords = "cast convert", BlueprintAutocast), Category = "Socket")
+	static TArray<uint8> Conv_DoubleToBytes(double InDouble)
```
Comment (CodeSpartan) [inline Source/TcpSocketPlugin/Public/TcpSocketConnection.h]: Please move the function definition into .cpp

File: Source/TcpSocketPlugin/Public/TcpSocketConnection.h [hunk 4]
```diff
@@ -72,34 +72,116 @@ class TCPSOCKETPLUGIN_API ATcpSocketConnection : public AActor
...
+	UFUNCTION(BlueprintCallable, meta = (DisplayName = "Read Bytes", Keywords = "read bytes"), Category = "Socket")
+	static bool Message_ReadBytes(int32 NumBytes, UPARAM(ref) TArray<uint8>& Message, TArray<uint8>& ReturnArray);
 
 	UFUNCTION(BlueprintCallable, meta = (DisplayName = "Read Byte", Keywords = "read byte int8 uint8"), Category = "Socket")
 	static uint8 Message_ReadByte(UPARAM(ref) TArray<uint8>& Message);
 
-	UFUNCTION(BlueprintCallable, meta = (DisplayName = "Read Bytes", Keywords = "read bytes"), Category = "Socket")
-	static bool Message_ReadBytes(int32 NumBytes, UPARAM(ref) TArray<uint8>& Message, TArray<uint8>& ReturnArray);
+	UFUNCTION(BlueprintCallable, meta = (DisplayName = "Read Short", Keywords = "read short int16"), Category = "Socket")
+	static int16 Message_ReadShort(UPARAM(ref) TArray<uint8>& Message)
```
Comment (CodeSpartan) [inline Source/TcpSocketPlugin/Public/TcpSocketConnection.h]: Please move the function definition into .cpp

File: Source/TcpSocketPlugin/Public/TcpSocketConnection.h [hunk 5]
```diff
@@ -72,34 +72,116 @@ class TCPSOCKETPLUGIN_API ATcpSocketConnection : public AActor
...
+		{
+			return 0;
+		}
+	}
+
+	UFUNCTION(BlueprintCallable, meta = (DisplayName = "Read Int", Keywords = "read int"), Category = "Socket")
+	static int32 Message_ReadInt(UPARAM(ref) TArray<uint8>& Message);
+
+	UFUNCTION(BlueprintCallable, meta = (DisplayName = "Read Long", Keywords = "read long int64"), Category = "Socket")
+	static int64 Message_ReadLong(UPARAM(ref) TArray<uint8>& Message)
```
Comment (CodeSpartan) [inline Source/TcpSocketPlugin/Public/TcpSocketConnection.h]: Please move the function definition into .cpp

File: Source/TcpSocketPlugin/Public/TcpSocketConnection.h [hunk 6]
```diff
@@ -72,34 +72,116 @@ class TCPSOCKETPLUGIN_API ATcpSocketConnection : public AActor
...
+		{
+			return 0;
+		}
+	}
 
 	UFUNCTION(BlueprintCallable, meta = (DisplayName = "Read Float", Keywords = "read float"), Category = "Socket")
 	static float Message_ReadFloat(UPARAM(ref) TArray<uint8>& Message);
 
+	UFUNCTION(BlueprintCallable, meta = (DisplayName = "Read Double", Keywords = "read double"), Category = "Socket")
+	static double Message_ReadDouble(UPARAM(ref) TArray<uint8>& Message)
```
Comment (CodeSpartan) [inline Source/TcpSocketPlugin/Public/TcpSocketConnection.h]: Please move the function definition into .cpp

**Review by CodeSpartan — CHANGES_REQUESTED** [review]

**Comment by PantelisAndrianakis** [comment]
I made three additional changes.
One for TcpSocketConnection.h removing function code from there.
And one for TcpSocketConnection.cpp moving conversion code there.
Also replaced sizeof with simple byte count to match the pre-existing Int method code.

**Comment by CodeSpartan** [comment]
Your PR doesn't compile: there's no `int16` type in blueprints, so it can't be exposed to blueprints. Leave the int16-related functions for C++ users, but don't expose them to blueprints. There's also some syntax errors.

In blueprints, Long is called `Integer 64`, so the blueprint names should conform. The double type in Blueprints is called a `Float (double-precision)`, so same thing here. The old `Read Float` should appear as `Read Float (single-precision)` and `Read Double` should appear as `Read Float (double-precision)`. Same with hover texts that are grabbed from the comments, like "Converts a float/double/long to an array of bytes" should feature new type names.
The C++ names should remain as they are now.

Please use in-place initialization of arrays, e.g.
```cpp
return TArray {
	static_cast<uint8>((InLong >> 56) & 0xFF),
	static_cast<uint8>((InLong >> 48) & 0xFF),
	//...
}
```
I know I've used `Array.Add()` in existing code, but those are previously made bad practices.

**Comment by PantelisAndrianakis** [comment]
int16-related functions are needed so that you use less bandwidth on network transfer (eg player inventory item infos).
Same can be done with booleans but is very rare, so I did not add that.
On my server side short values are return as full int32 as well.
So I made the function to return an int32 as well.

I do not agree with naming conformations you proposed, Long and Double are self explainable.
At the end of the day, it is a matter of taste, could also use D for int, L for long etc, but that would make them harder to read.

Used in-place initialization at Conv_LongToBytes.

**Comment by CodeSpartan** [comment]
Merging your PR into a separate branch to align with Epic's naming standards and ensure functionality. Will make necessary adjustments before merging into main. Thank you for your contribution.

**Pull request merged**